fraud/bad_encoding.go

```diff
@@ -0,0 +1,219 @@
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// Shares contains all shares from row/col
+	// Shares that did not pass verification in rmst2d will be nil
+	// For non-nil shares MerkleProofs are computed
+	Shares []*Share
+	// Position represents the number of row/col where ErrByzantineRow/ErrByzantineColl occurred
+	Position uint8
+	isRow    bool
+}
+
+func CreateBadEncodingFraudProof(
+	ctx context.Context,
+	height uint64,
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	roots [][]byte,
+	errShares [][]byte,
+	dag format.NodeGetter,
+) (Proof, error) {
+	shares := make([]*Share, len(errShares))
+	for index, share := range errShares {
+		if share != nil {
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+
+			tree, err := buildTreeFromLeaves(ctx, data, roots[index], index, dag)
+			if err != nil {
+				return nil, err
+			}
+			proof, err := tree.Tree().Prove(int(position))
+			if err != nil {
+				return nil, err
+			}
+			/*
+				NamespaceID should be replaced with ParitySharesNamespaceID for
+				all shares except Q0(all erasure coded shares).
+				For EDS 4x4 Q0 is [[0;0],[0;1],[1:0],[1;1]].
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
+			if index >= len(errShares)/2 || int(position) >= len(errShares)/2 {
+				namespaceID = consts.ParitySharesNamespaceID
+			}
+
+			shares[index] = &Share{namespaceID, share, &proof}
+		}
+	}
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares,
+		isRow:       isRow,
+		Position:    position,
+	}, nil
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to []byte
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEnconding{
+		Height:   p.BlockHeight,
+		Shares:   shares,
+		Position: uint32(p.Position),
+		IsRow:    p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
+	in := pb.BadEnconding{}
+	if err := in.Unmarshal(data); err != nil {
+		return err
+	}
+	befp := UnmarshalBefp(&in)
+	*p = *befp
+
+	return nil
+}
+
+func UnmarshalBefp(befp *pb.BadEnconding) *BadEncodingProof {
+	return &BadEncodingProof{
+		BlockHeight: befp.Height,
+		Shares:      ProtoToShare(befp.Shares),
+	}
+}
+
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader, codec rsmt2d.Codec) error {
```

Comment:
Pls use the codec constructor defined in [core/consts](https://github.com/celestiaorg/celestia-core/blob/master/pkg/consts/consts.go#L75)

Reply:
And not just here, but every where you create the codec.